Leap Seconds Support

tarekgh opened the pull request:
Port the leap seconds changes done in the full framework to net core.

tarekgh commented:
@jkotas this is just porting what we have done in the full framework. Also, I have created a commit for corert we can use when we are mirroring this change there https://github.com/tarekgh/corert/commit/f79f7c8b348e38378b2a4b2566fafe53172ae250

could you please take a look at both and let me know if you have any comment? I have tested this on coreclr, corert and ProjectN native and Win32. I did the testing on Windows version which support leaps seconds with inserting some fake leap seconds to the system.

jkotas commented on src/System.Private.CoreLib/src/System/DateTime.Unix.cs:
```diff
@@ -0,0 +1,54 @@
+{
+    public readonly partial struct DateTime
+    {
+        internal static readonly bool s_isLeapSecondsSupportedSystem = false;
```
Can this be const on Unix?

jkotas commented on src/classlibnative/bcltype/system.cpp:
```diff
@@ -99,12 +99,142 @@ FCIMPL0(INT64, SystemNative::__GetSystemTimeAsFileTime)
+    BEGIN_QCALL;
+
+    SYSTEM_LEAP_SECOND_INFORMATION slsi;
+    HINSTANCE hNtDll = CLRLoadLibrary(W("ntdll.dll"));
```
Can we just PInvoke into `ntdll.lib` all in C# like you have done it for CoreRT version of this change?

tarekgh replied:
I was not sure about if there will be any perf difference. corert I didn't have the choice. if you think perf wil not be issue I can convert this to pinvoke.

jkotas replied:
The one-time initialization is certainly not a performance sensitive path.

jkotas replied:
(The rest can stay as FCalls for now. There are probably a few perf sensitive ones.)

jkotas commented on src/classlibnative/bcltype/system.cpp:
```diff
@@ -99,12 +99,142 @@ FCIMPL0(INT64, SystemNative::__GetSystemTimeAsFileTime)
+    if (::FileTimeToSystemTime((FILETIME*)&fileTime, (LPSYSTEMTIME) time))
+    {
+
+#if BIGENDIAN
```
You can delete `BIGENDIAN` ifdefs in Windows-specific code. Windows is never going to run on big endian processors.

jkotas commented on src/System.Private.CoreLib/src/System/DateTime.Windows.cs:
```diff
@@ -0,0 +1,182 @@
+                if (s_isLeapSecondsSupportedSystem)
+                {
+                    FullSystemTime time = new FullSystemTime();
+                    GetSystemTimeWithLeapSecondsHandling(ref time);
```
Can this argument be `out` to avoid the unnecessary upfront zero-initialization?

jkotas commented on src/System.Private.CoreLib/src/System/DateTime.Windows.cs:
```diff
@@ -0,0 +1,182 @@
+            throw new ArgumentOutOfRangeException(null, SR.ArgumentOutOfRange_FileTimeInvalid);
+        }
+
+        // Just in case for any reason CreateDateTimeFromSystemTime not get inlined,
```
You can pass it as `in` to make the intent more obvious

jkotas commented on src/System.Private.CoreLib/src/System/DateTime.Unix.cs:
```diff
@@ -0,0 +1,54 @@
+        {
+            get
+            {
+                // following code is tuned for speed. Don't change it without running benchmark.
```
This comment and the structure of the code does not really make sense - I guess it is a copy&paste from somewhere.

I would change it to:

`return new DateTime(((ulong)(GetSystemTimeAsFileTime() + FileTimeOffset)) | KindUtc);`

tarekgh replied:
Yes this is copy and paste from DateTime.CoreClr.cs. but I'll change it as you suggested

jkotas commented on src/System.Private.CoreLib/src/System/DateTime.Windows.cs:
```diff
@@ -0,0 +1,182 @@
+        {
+            get
+            {
+                // following code is tuned for speed. Don't change it without running benchmark.
```
Sam here

tarekgh replied:
ditto

jkotas commented on src/classlibnative/bcltype/system.h:
```diff
@@ -38,6 +51,13 @@ class SystemNative
+    static FCDECL1(VOID, GetSystemTimeWithLeapSecondsHandling, FullSystemTime *time);
+    static FCDECL2(FC_BOOL_RET, ValidateSystemTime, FullSystemTime *time, CLR_BOOL localTime);
+    static FCDECL2(FC_BOOL_RET, SystemFileTimeToSystemTime, INT64 fileTime, FullSystemTime *time);
+    static FCDECL2(FC_BOOL_RET, SystemNative::SystemTimeToSystemFileTime, FullSystemTime *time, INT64 *pFileTime);
```
` SystemNative::` is not necessary here. (And I believe it will cause warning/error with some compilers.)

jkotas commented:
Is there going to be a matching PR with the tests in CoreFX?

jkotas commented on src/System.Private.CoreLib/src/System/DateTime.Windows.cs:
```diff
@@ -0,0 +1,182 @@
+                hundredNanoSecond = 0;
+            }
+
+            internal ushort wYear;
```
Would it look better to have actual SYSTEMTIME field here? I looks like it would save us some casts where we pass around `FullSystemTime`, but we are actually passing just `SYSTEMTIME`.

tarekgh replied:
will do that.

tarekgh commented:
> Is there going to be a matching PR with the tests in CoreFX?

We already have tests for DateTime/DateTimeOffset exercise Now, UtcNow. Writing any test specific for testing with leap seconds will not have any effect because we'll have to run on a machine support leap seconds (i.e. RS5) and the system will have a leaps seconds which we don't have any so far. This will be interesting more when we get in the future a leap seconds.

jkotas commented:
> We already have tests

There are two separate problems:
1. Where/how/when we have machines with the right config available in automation. I understand that this may be not easy to solve momentarily. I assume that you got the right machine yourself. 
2. Tests that hit the new code on the machine with the right config. These tests should be failing before your change and passing after your change. This is the part that I think should be send to CoreFX.

tarekgh commented:
> Tests that hit the new code on the machine with the right config. These tests should be failing before your change and passing after your change. This is the part that I think should be send to CoreFX.

I'll write a test for that but I'll not expect this test will fail now and the possibility to fail in the future will be low. I understand your point and would be good to have it to catch if we would have a problem.

danmosemsft commented on src/System.Private.CoreLib/src/System/DateTime.Windows.cs:
```diff
@@ -0,0 +1,182 @@
+{
+    public readonly partial struct DateTime
+    {
+        internal static readonly bool s_isLeapSecondsSupportedSystem = IsLeapSecondsSupportedSystem();
```
`isLeapSecondsSupportedSystem` does not sound right to me. What about `s_systemSupportsLeapSeconds` ? 

Then code reads better:
```c#
            if (s_systemSupportsLeapSeconds)
            {
                return InternalFromFileTime(fileTime);
            }
```

danmosemsft commented on src/System.Private.CoreLib/src/System/DateTime.Windows.cs:
```diff
@@ -0,0 +1,182 @@
+            return new DateTime( ((UInt64)(ticks)) | KindUtc);
+        }
+
+        // FullSystemTime struct matches Windows SYSTEMTIME struct, except we added the extra nanoSeconds field to store
```
nanoSeconds - > hundredNanoSecond  ?

jkotas commented on src/System.Private.CoreLib/System.Private.CoreLib.csproj:
```diff
@@ -432,6 +432,7 @@
   </ItemGroup>
   <ItemGroup Condition="'$(TargetsWindows)' == 'true'">
     <Compile Include="$(BclSourcesRoot)\System\DateTime.Windows.cs" />
+    <Compile Include="shared/Interop/Windows/NtDll/Interop.NtQuerySystemInformation.cs" />
```
This should be in the project in the shared partition

jkotas commented on src/System.Private.CoreLib/src/System/DateTime.Windows.cs:
```diff
@@ -139,42 +142,32 @@ internal FullSystemTime(long ticks)
 
         [MethodImplAttribute(MethodImplOptions.InternalCall)]
-        internal static extern bool SystemTimeToSystemFileTime(ref FullSystemTime time, ref long fileTime);
+        internal static extern bool SystemTimeToSystemFileTime(ref Interop.Kernel32.SYSTEMTIME time, out long fileTime);
```
The first argument can be `in Interop.Kernel32.SYSTEMTIME`

jkotas commented on src/System.Private.CoreLib/src/System/DateTime.Windows.cs:
```diff
@@ -139,42 +142,32 @@ internal FullSystemTime(long ticks)
-
-        [DllImport(JitHelpers.QCall, CharSet = CharSet.Unicode)]
-        internal static extern bool IsLeapSecondsSupportedSystem();
+        internal static extern bool ValidateSystemTime(ref Interop.Kernel32.SYSTEMTIME time, bool localTime);
```
The first argument can be `in`

jkotas commented on src/System.Private.CoreLib/src/System/DateTime.Windows.cs:
```diff
@@ -71,64 +67,71 @@ internal static bool IsValidTimeWithLeapSeconds(int year, int month, int day, in
-            FullSystemTime time = new FullSystemTime();
-            if (SystemFileTimeToSystemTime(fileTime, ref time))
+            FullSystemTime time;
+            if (SystemFileTimeToSystemTime(fileTime, out time))
```
Nit: `if (SystemFileTimeToSystemTime(fileTime, out FullSystemTime time))`

jkotas commented on src/System.Private.CoreLib/src/System/DateTime.Windows.cs:
```diff
@@ -71,64 +67,71 @@ internal static bool IsValidTimeWithLeapSeconds(int year, int month, int day, in
+            long fileTime;
             FullSystemTime time = new FullSystemTime(ticks);
-            if (SystemTimeToSystemFileTime(ref time, ref fileTime))
+            if (SystemTimeToSystemFileTime(ref time.systemTime, out fileTime))
```
Nit: `if (SystemTimeToSystemFileTime(ref time.systemTime, out long fileTime))`

jkotas commented on src/System.Private.CoreLib/src/System/DateTime.Windows.cs:
```diff
@@ -0,0 +1,175 @@
+        internal static extern bool ValidateSystemTime(ref Interop.Kernel32.SYSTEMTIME time, bool localTime);
+
+        [MethodImplAttribute(MethodImplOptions.InternalCall)]
+        internal static extern bool SystemFileTimeToSystemTime(long fileTime, out FullSystemTime time);
```
Is there a reason why we call some of these things `SystemFileTime`? There is `SYSTEMTIME` and `FILETIME` today, but there is no SystemFileTime today.

jkotas commented on src/System.Private.CoreLib/src/System/DateTime.Windows.cs:
```diff
@@ -0,0 +1,175 @@
+            }
+
+            ticks -= FileTimeOffset;
+            if (ticks < 0) {
```
Same

tarekgh commented on src/System.Private.CoreLib/src/System/DateTime.Windows.cs:
```diff
@@ -0,0 +1,175 @@
+            return new DateTime(universalTicks, DateTimeKind.Utc);
+        }
+
+        public long ToFileTimeUtc() {
```
Those was a copy and paste from other file. I'll fix them.

jkotas commented on src/System.Private.CoreLib/src/System/DateTime.Windows.cs:
```diff
@@ -0,0 +1,175 @@
+                hundredNanoSecond = 0;
+            }
+
+            internal Interop.Kernel32.SYSTEMTIME systemTime;
```
The coding convention is to place the field first.

jkotas commented on src/System.Private.CoreLib/src/System/DateTime.Windows.cs:
```diff
@@ -0,0 +1,175 @@
+            }
+        }
+
+        public static DateTime FromFileTimeUtc(long fileTime)
```
The only difference between Windows and Unix versions of FromFileTimeUtc and ToFileTimeUtc is the callout to the special method method for leap seconds. Should we have keep it in the shared and have a dummy method on Unix?

Also, it may be better to call InternalFromFileTime / InternalToFileTime something like `FromFileTimeLeapSecondsAware / ToFileTimeLeapSecondAware`.

jkotas commented:
Do you have the number for performance hit that the methods like `DateTime.UtcNow` will incur on the leap second aware systems with this change?

I think we had it in one of the offline conversations ... it would be useful to have the number on here too.

jkotas commented:
Looks good to me otherwise

tarekgh commented:
> Do you have the number for performance hit that the methods like DateTime.UtcNow will incur on the leap second aware systems with this change?

I don't have the exact numbers here and also I am using a VM for my leap seconds testing. but I have tested it and we always reporting the time on the same milliseconds boundary. Here example for what I was doing:

```C#
            SYSTEMTIME st = new SYSTEMTIME();
            GetSystemTime(ref st);
            DateTime dt = DateTime.UtcNow;
            Console.WriteLine($"<{st.wHour:D2}:{st.wMinute:D2}:{st.wSecond:D2}.{st.wMillisecond:D2}> .... [{dt.Hour:D2}:{dt.Minute:D2}:{dt.Second:D2}.{dt.Millisecond:D2}]");
```

Note that we still reporting the precise time as we used to do even in leap seconds system. the only extra perf is just the time we use till we return. I can do the measurements later to get the exact numbers. note that we already have a perf tests for this too. anyway, let me know if you want to wait not merging the change till we do that or we case merge and do the measurements later. if we want to wait, I'll hold this changes till next year.

jkotas commented:
>  let me know if you want to wait not merging the change till we do that

I do not think you need to block on this.

tarekgh commented:
test Windows_NT x64 Checked Innerloop Build and Test (Jit - TieredCompilation=0)

tarekgh commented:
thanks @jkotas for your review and comments. I will wait preparing the changes for corert too and test it so we'll be ready when we merge. I'll let you know when I have it ready to take a quick look there too.

tarekgh commented:
here is the delta changes in corert repo too https://github.com/tarekgh/corert/commit/b5037421cc746961f6484f5c7b164405dfa7365f i have tested it with PN too. after reviewing the delta, I'll squash moth commits in corert to one so it will be easy to use.

tarekgh commented:
Thanks a lot @jkotas for your review. I have updated the corert too and squashed the commits into one commit https://github.com/tarekgh/corert/commit/389615a4b5528321b945a2dc0400ede4256a031c

tarekgh commented:
I added a comment too on corert change https://github.com/tarekgh/corert/commit/389615a4b5528321b945a2dc0400ede4256a031c#r31609074 may be we need to apply.

tarekgh commented:
@Anipik I am noticing there is no corert mirroring PR opened for this one. Is it expected?

Anipik commented:
@tarekgh  it is expected in the sense that there can be only one open mirror pr in a repo at any time.
And we have already have one in corert which hasnt yet merged https://github.com/dotnet/corert/pull/6663
After this PR has been merged the next pr for corert repo will contain this commit
